src/codegen/traverse.lisp

```diff
@@ -253,6 +254,60 @@ other nodes, then it would be inappropriate to also define an action
       (let ((*traverse* #'current-traverse))
         (apply *traverse* initial-node initial-args)))))
 
+(defun traverse-with-path (node action-list &rest args)
+  "Like 'traverse', but actions receive a thunk that returns a reverse
+list of the current node's ascendants.
+That is, `car` of the ascendants is the node's immediate parent,
+and the last element of ascendants is the root of the AST on which
+`traverse-with-path` is called.  If visiting node is the root, the
+ascendant list is empty."
+  (declare (type node node)
+           (values node &optional))
+  (let ((traversal-path nil))
+    (labels ((wrap-action (when action)
+               (if action
+                   (ecase when
+                     (:before
+                      (make-action when 'node
+                                   (lambda (node path-thunk &rest args)
+                                     (push node traversal-path)
+                                     (apply (action-function action) node path-thunk args))))
+                     (:after
+                      (make-action when 'node
+                                   (lambda (node path-thunk &rest args)
+                                     (let ((r (apply (action-function action)
+                                                     node
+                                                     path-thunk
+                                                     args)))
+                                       (pop traversal-path)
+                                       r)))))
+                   (make-action when 'node
+                                (lambda (node &rest _rest)
+                                  (declare (ignore rest))
+                                  (funcall op node))))))
```

Comment:
Wait, what is `op`? Is this supposed to be a copy of the `ecase` above, just with the `(apply (action-function action) node path-thunk args)` removed?

Reply:
Good catch.  Remnant of the previous version.